### ISSUE TITLE
Migration from pq to pgx

### ISSUE THREAD

**Issue by krzysztofdrys** [jackc/pgx#104]
We are considering migration from pq to pgx. After creating proof of concept we have found the following problems:

- where  pq would send `[]byte`  to `Scan` method, pgx sends  `string`. We encountered this issue with `TEXT` column, but also `JSONB` and few other column types. Why is it so?
- We have a query which looks something like this `SELECT * FROM foo where EXTRACT('EPOCH' FROM created_on) >= :from`. pq accepts `int64` as `from` argument. pgx requires `from` to be `float64`. Is it desired?

**Comment by jackc** [comment]
Are you using the native pgx interface or as a driver for database/sql? I'll answer assuming the native interface.

pgx supports scanning any data type into `[]byte`. See https://github.com/jackc/tpr/blob/master/backend/pgx_repository.go#L153 for an example of reading a `json` type into a `[]byte` (The source for the prepared statement is at https://github.com/jackc/tpr/blob/master/backend/pgx_repository.go#L425)

As far as the `float64` vs. `int64` in the extract query, the `epoch` expression is a PostgreSQL `double precision` value.

```
jack=# select pg_typeof(extract('epoch' from now())), extract('epoch' from now());
    pg_typeof     |    date_part
------------------+------------------
 double precision | 1447201631.16095
(1 row)

Time: 0.542 ms
```

pgx internally creates a prepared statement for all queries that have result sets. The type of the bound variable will also be a `double precision` as that is what it is being compared to. pgx requires the Go types to exactly match the PostgreSQL types (with a few limited exceptions). If you really want to pass in a Go int64 to compare with a PostgreSQL `double precision` you could use a type cast in your SQL.

```sql
SELECT * FROM foo where EXTRACT('EPOCH' FROM created_on) >= $1::int64
```

**Comment by krzysztofdrys** [comment]
Thanks for your answer. 

Regarding `float64` vs. `int64`, I guess that everything is clear now.

I am still not sure about using `pgx` as a `sql` driver. So yes, at first we want to use `pgx` as a `sql` driver - sorry about not being clear about this in my first post.

I will try to summarize our problem and our set-up. We keep info about links to social sites in the following structure:
```
type Links []string

func (l *Links) Scan(value interface{}) error {
	return json.Unmarshal([]byte(value.(string)), l)
}
```

In database this information is stored inside `jsonb` column. There are two notable differences between `pq` and `pgx`. 

First is that `value` used to be `[]byte` in `pq`. So our `Scan` function used to look like this: return `json.Unmarshal(value.([]byte), l)`. What used to be a type assertion, now is a type assertion and conversion from `string` to `[]byte`. Performance cost is probably not that big, but still I would prefer to avoid it.

The second problem is more painful. It occurs, when `null` is stored is db. And I don't mean db `null`, I mean json `null`. To understand the difference, please inspect the following two `sql` queries:

```
-- inserts db null
update users
set links = null
where id = 101;

-- inserts json null
update users
set links = 'null'
where id = 101;
```

So, when json null is stored in db, `Scan` method receives string `"bnVsbA=="`. The same string can be obtained from the following code: 
```
b, err := json.Marshal([]byte("null"))
fmt.Println(string(b))
```
so I suspect that `pgx` tries to perform some marshaling/unmarshaling on the value it receives from the db.

Anyway, `"bnVsbA=="` cannot be properly unmarshaled into `Links` and scan fails. This of  course is undesired. With `pq` we received `"null"`, which was properly unmarshaled into (empty) `Links` object.

TL;DR When scanning (using `pgx` as a driver) `jsonb` value we receive `"bnVsbA=="` instead of `"null"`. What can we do about it? Is this desired?

**Comment by jackc** [comment]
Can you try master? I just made a tweak that makes `[]byte` values to skip the normal encoding/decoding process. The problem with JSON was because pgx automatically marshals and unmarshals values going into or out of a PostgreSQL `json` or `jsonb` type. base64 is what `json.Marshal` does to a `[]byte`.

**Comment by krzysztofdrys** [comment]
I have just checked and the issue is solved at master. Thanks for your help!
